File: benchmarks/crivo.c
#include <stdio.h>
#include <stdlib.h>

#define N 500000

int main() {
    int* crivo = (int*)malloc((N + 1) * sizeof(int));
    for (int i = 0; i <= N; i++) crivo[i] = 1;
    crivo[0] = 0;
    crivo[1] = 0;

    for (int i = 2; i * i <= N; i++) {
        if (crivo[i]) {
            for (int j = i * i; j <= N; j += i) {
                crivo[j] = 0;
            }
        }
    }

    int contagem = 0;
    for (int i = 2; i <= N; i++) {
        contagem += crivo[i];
    }

    printf("%d\n", contagem);
    free(crivo);
    return 0;
}

File: benchmarks/crivo.rs
const N: usize = 1_000_000;

fn main() {
    let mut crivo = vec![1i32; N + 1];
    crivo[0] = 0;
    crivo[1] = 0;

    let mut i = 2;
    while i * i <= N {
        if crivo[i] == 1 {
            let mut j = i * i;
            while j <= N {
                crivo[j] = 0;
                j += i;
            }
        }
        i += 1;
    }

    let mut contagem = 0i32;
    for k in 2..=N {
        contagem += crivo[k];
    }

    println!("{}", contagem);
}
